components/raftstore/src/store/peer.rs

```diff
@@ -3501,11 +3514,32 @@ where
         Ok((min_m, min_c))
     }
 
-    fn pre_propose_prepare_merge<T>(
-        &self,
+    fn pre_propose_prepare_merge<T: Transport>(
```

Comment:
Seems here is the only place (except leader change) that may update `prepare_merge_fence`, does it mean we very rely on PD to retry the merge command to push forward the merge process? If the merge request is keep failing (perhaps due to the `pessimistic_locks` is too large) and PD end the merge operator with a timeout error hence won't retry the merge command anymore, the region may remain in merging mode forever and won't accept incoming proposals.

Reply:
It is also retried in `on_apply_res`. Which means, when the `applied_index` reaches `prepare_merge_fence`, we will continue.

Reply:
In `on_apply_res`, we retried with the previous saved `PrepareMerge` request, it is also required the request can still pass all the checks before `pre_propose_prepare_merge`

Reply:
Sounds like something worth considering. But what check may it fail to pass before `pre_propose_prepare_merge` while there is not any region change and the region can still serve request?

Reply:
I have been checked the code, checks that rely on states that outside of the source peer, like the disk full check and the target peer check:
https://github.com/tikv/tikv/blob/4e2af0426c13657c6eeb69bffc7b9847f2f61b47/components/raftstore/src/store/fsm/peer.rs#L3886-L3902

which may change after waiting for the source peer's `appied_index` catch up. 

The source peer's state can also change as apply more log, but with the `epoch_checker` there won't have `EpochNotMatch` error though.

Reply:
Now I always clear the fence after proposing `PrepareMerge` in `on_apply_res`. I think the region will return available even if `PrepareMerge` fails to propose.